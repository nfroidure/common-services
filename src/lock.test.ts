import assert from 'assert';
import sinon from 'sinon';
import Knifecycle, { constant } from 'knifecycle';
import initLockService from './lock';
import type { DelayService } from './delay';

describe('initLockService', () => {
  const log = sinon.stub();
  const delay: DelayService = {
    create: sinon.stub(),
    clear: sinon.stub(),
  };

  beforeEach(() => {
    log.reset();
  });

  test('should work', async () => {
    const lock = await initLockService({
      log,
      delay,
    });

    assert('object' === typeof lock);
    assert.deepEqual(log.args, [['debug', 'ðŸ”’ - Lock service initialized.']]);

    await lock.take('key');
    await lock.take('key2');
    lock.release('key');
    lock.release('key2');

    await Promise.all([
      lock.take('key').then(() => lock.release('key')),
      lock.take('key').then(() => lock.release('key')),
      lock.take('key').then(() => lock.release('key')),
      lock.take('key').then(() => lock.release('key')),
    ]);

    expect({
      logCalls: log.args,
    }).toMatchSnapshot();
  });

  test('should work with Knifecycle', (done) => {
    new Knifecycle()
      .register(initLockService)
      .register(constant('log', log))
      .register(constant('delay', delay))
      .run(['lock'])
      .then(({ lock }) => {
        assert(lock);
        assert.deepEqual(log.args, [
          ['debug', 'ðŸ”’ - Lock service initialized.'],
        ]);
      })
      .then(() => done())
      .catch(done);
  });
});
