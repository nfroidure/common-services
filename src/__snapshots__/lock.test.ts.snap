// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initLockService should work 1`] = `
Object {
  "logCalls": Array [
    Array [
      "debug",
      "🔒 - Lock service initialized.",
    ],
    Array [
      "debug",
      "🔐 - Taking the lock on key (queue length was 0)",
    ],
    Array [
      "debug",
      "🔐 - Taking the lock on key2 (queue length was 0)",
    ],
    Array [
      "debug",
      "🔓 - Releasing the lock on key (queue length was 1)",
    ],
    Array [
      "debug",
      "🔓 - Releasing the lock on key2 (queue length was 1)",
    ],
    Array [
      "debug",
      "🔐 - Taking the lock on key (queue length was 0)",
    ],
    Array [
      "debug",
      "🔐 - Taking the lock on key (queue length was 1)",
    ],
    Array [
      "debug",
      "🔐 - Taking the lock on key (queue length was 2)",
    ],
    Array [
      "debug",
      "🔐 - Taking the lock on key (queue length was 3)",
    ],
    Array [
      "debug",
      "🔓 - Releasing the lock on key (queue length was 4)",
    ],
    Array [
      "debug",
      "🔓 - Releasing the lock on key (queue length was 3)",
    ],
    Array [
      "debug",
      "🔓 - Releasing the lock on key (queue length was 2)",
    ],
    Array [
      "debug",
      "🔓 - Releasing the lock on key (queue length was 1)",
    ],
  ],
}
`;
